### DISCUSSION TITLE
Parameterized Tests - Find & Run Independently

### DISCUSSION THREAD

**Discussion by DirttDiver** [xunit/xunit#2267] [Question] [answered]
I have been using Xunit for 2 months now (C#) and I have come from a pytest background (Python) 
I have been trying to parametrize tests using properties set in a class and have used the following from various stackoverflow sources

         public static IEnumerable<object[]> Navigation_Params()
        {
            NavigationHolder navy = new NavigationHolder();
            

    
            yield return new object[] { navy.pip_builder.dash_expand_pip_builder_menu, navy.pip_builder.dash_work_stages, "PROJECT STAGES" };
            yield return new object[] { navy.pip_builder.dash_expand_pip_builder_menu, navy.pip_builder.dash_design_disciplines, "PROJECT DESIGN DISCIPLINES" };
            //yield return new object[] { issues.snag_radial };
        }

```
[Theory]
        [MemberData(nameof(Data))]
        public void Test_Navigation_Menus(By menu_section, By sub_section, string page_heading_title)
        {
            nav_met.Verify_Navigation_Menus(menu_section, sub_section, page_heading_title);
        }
```

This works fine, however I have 2 issues. 

- The test discovery considers the test as 1 test and not the amount of objects passed in e.g. if i had 5 objects, I would expect for the test discovery to display 5 separate tests. (the inline data decorators works perfectly for this, however you can't use properties as a parameter.
- Using my code above, the tests do run, but they run in series, it would be extremely beneficial to be able to run these in parallel 

My question, is it possible to achieve what I have proposed with the current version of Xunit or is there plans to implement anything along the lines of what I have mentioned?

Thanks!

**Comment by bradwilson** [comment, marked as answer]
>The test discovery considers the test as 1 test and not the amount of objects passed in e.g. if i had 5 objects, I would expect for the test discovery to display 5 separate tests. (the inline data decorators works perfectly for this, however you can't use properties as a parameter.

This is...sort of true, sometimes. :) We have a separate concept of test case vs. test. A test case results in one or more tests being run, each with an individual result. For `[Fact]` tests, there is always a 1:1 mapping between test case and test. For `[Theory]` tests, there two situations which could arise: for some situations, you get a 1:1 mapping between test case and test (where each row of data is a test case), and in some situations you get a 1:* mapping between test case and test (where the test method is the test case, and each row of data is a test).

What determines this "sometimes" situation? Not to put too fine point on it, but: Visual Studio's Test Explorer. In this case, because of the design of Test Explorer (and the underlying engine, really) we need to be able to serialize test cases. If the theory data is not serializable, then we have to fall back to the 1:* mapping scenario where we can serialize information about the test method itself and call it a single test case, vs. if we can serialize all the theory data we can use the 1:1 mapping scenario.

(There is a [documentation page](https://xunit.net/faq/theory-data-stability-in-vs) on "Theory Data Stability", which is not your exact problem but does describe the serialization process a bit, that might be useful as additional background information.)

By your description, I'm going to guess you're running these tests in Visual Studio. When that's the case, we need to be able to serialize your theory data, which I'm guessing we cannot. So we fall back to the 1:* mapping. This means the Visual Studio UI will only show you the single test method.

Here's an example which has a test where all the data is serializable vs. not:

![image](https://user-images.githubusercontent.com/16944/109401344-f8714380-7902-11eb-8c27-a2bf8fc2ecf5.png)

You can see that the test with serializable data shows individually runnable and reportable results, vs. the one that does not:

![image](https://user-images.githubusercontent.com/16944/109401378-2b1b3c00-7903-11eb-82d3-795faddce939.png)

![image](https://user-images.githubusercontent.com/16944/109401390-3a9a8500-7903-11eb-99f7-8d7d79965919.png)

So, what can you do to make your data serializable? This is the "rule set" we use, which basically includes a bunch of intrinsic types + anything which implements `IXunitSerializable`: https://github.com/xunit/xunit/blob/a8614d34999c889e1c8014f679de95d20eae1304/src/common/XunitSerializationInfo.cs#L353-L388

Note that this custom serialization type has been removed in v3 (not yet released) in favor of built-in serialization. We couldn't use built-in serialization at the time because our target platforms didn't all support it.

>Using my code above, the tests do run, but they run in series, it would be extremely beneficial to be able to run these in parallel

This isn't possible today in v2, but it's a feature we're working on for v3. Our [road map](https://github.com/xunit/xunit/issues/2133) contains two issues in particular: https://github.com/xunit/xunit/issues/1986 https://github.com/xunit/xunit/issues/2055
